subsys/bluetooth/host/l2cap.c

```diff
@@ -1954,7 +1955,8 @@ static int l2cap_chan_le_send(struct bt_l2cap_le_chan *ch,
 	/* Save state so it can be restored if we failed to send */
 	net_buf_simple_save(&buf->b, &state);
 
-	if (net_buf_frags_len(buf) <= ch->tx.mps) {
+	if ((buf->len <= ch->tx.mps) &&
+	    (net_buf_headroom(buf) >= BT_L2CAP_BUF_SIZE(0))) {
```

Comment:
Is this ensured by a check somewhere else now? Can it be an assert?

Reply:
No.
The problem is that here, `buf` can be any `buf` in a chain of fragments.
If we have fragments allocated e.g. by the networking subsys, the first one will be checked by the API entrypoint (`bt_l2cap_chan_send()`) but the rest of the chain won't.

I just asked the network subsys maintainer if they had a way for tweaking that, and if they do, I'd rather make a separate PR to fix the headroom req in the net subsys, and also remove the extra segmenting that happens here if that headroom req is not met.

Reply:
Update:
@rlubos says that there is no easy way to add a `reserve` to netbufs allocated by/for the net subsys.
The current way they do prepending is by allocating and prepending a whole net buf to the chain.

We might be able to optimize this (later) by keeping a ref on the previous `buf` and using that as our header space. But that requires the whole stack to be "frag-native" (TM) and adds a whole nother level of complexity that we don't need right now.

IMO since IPSP has been broken for a while and we didn't get much complaining, we should wait until someone really makes some noise about its perf/mem usage before attempting something like this.

Reply:
resolved IRL